Parsers/ASimDns/Parsers/vimDnsDnsMonster.yaml

```diff
@@ -0,0 +1,97 @@
+Parser:
+  Title: DNS activity ASIM filtering parser for DNSmonster
+  Version: '0.1'
+  LastUpdated: Mar 11 2022
+Product:
+  Name: DNSmonster
+Normalization:
+  Schema: Dns
+  Version: '0.1.3'
+References:
+- Title: ASIM DNS Schema
+  Link: https://aka.ms/ASimDnsDoc
+- Title: ASIM
+  Link: https://aka.ms/AboutASIM
+Description: |
+  This ASIM parser supports filtering and normalizing DNSmonster logs to the ASIM DNS activity normalized schema.
+ParserName: vimDnsDnsMonster
+ParserParams:
+  - Name: starttime
+    Type: datetime
+    Default: datetime(null)
+  - Name: endtime
+    Type: datetime
+    Default: datetime(null)
+  - Name: srcipaddr
+    Type: string
+    Default: '*'
+  - Name: domain_has_any
+    Type: dynamic
+    Default: dynamic([])
+  - Name: responsecodename
+    Type: string
+    Default: '*'
+  - Name: response_has_ipv4
+    Type: string
+    Default: '*'
+  - Name: response_has_any_prefix
+    Type: dynamic
+    Default: dynamic([])
+  - Name: eventtype
+    Type: string
+    Default: 'Query'
+
+ParserQuery: |
+ let rcodes = dynamic(["NOERROR","FORMERR","SERVFAIL","NXDOMAIN","NOTIMP","REFUSED","YXDOMAIN","YXRRSET","NXRRSET","NOTAUTH","NOTZONE","DSOTYPENI"]);
+ let rrcodes = dynamic(["RESV","A","NS","MD","MF","CNAME","SOA","MB","MG","MR","NULL","WKS","PTR","HINFO","MINFO","MX","TXT","RP","AFSDB","X25","ISDN","RT","NSAP","NSAP-PTR","SIG","KEY","PX","GPOS","AAAA","LOC","NXT","EID","NIMLOC","SRV","ATMA","NAPTR","KX","CERT","A6","DNAME","SINK","OPT","APL","DS","SSHFP","IPSECKEY","RRSIG","NSEC","DNSKEY","DHCID","NSEC3","NSEC3PARAM","TLSA","SMIMEA","UNASSIGN","HIP","NINFO","RKEY","TALINK","CDS","CDNSKEY","OPENPGPKEY","CSYNC","ZONEMD","SVCB","HTTPS"]);
+ let opcodes = dynamic(["Query","IQuery","Status","Unassigned","Notify","Update","DNS Stateful Operations"]);
+ let parser=
+    (
+    starttime: datetime=datetime(null), 
+    endtime: datetime=datetime(null),
+    srcipaddr: string='*',
+    domain_has_any: dynamic=dynamic([]),
+    responsecodename: string='*', 
+    response_has_ipv4: string='*',
+    response_has_any_prefix: dynamic=dynamic([]),
+    eventtype: string='Query',
+    disabled:bool=false
+    ) {
+    dnsmonster_CL | where not(disabled)
+    | where DNS_Response_b
+    | extend DNS_Question = todynamic(DNS_Question_s), DNS_Answer = todynamic(DNS_Answer_s)
+    | mv-expand DNS_Question
+    | mv-expand DNS_Answer
```

Comment:
Since the event includes a list of questions (though I suspect only one) and a list of answers, this dual mv-expand will create a cartesian multiplication which is not what you wanted. While for question, our analysis is that it is always one, for the answer, we just handle it as a single string in DnsResponseName.

Reply:
As noted above, I can remove the mv-expand for DNS_Question easily enough.  The DNS_Answer does routinely contain multiple values.  Is there a problem with the DNS_Answer expansion in order to perform the has_ipv4_prefix and ipv4_is_in_range tests against each result?

Reply:
Our solution to balance performance and usability was to leave the answer field unparsed and to use the ipv4 functions on the unparsed field. They work well.  

Few side notes:
- If you want to handle a JSON array, but not duplicate events, look into mv-apply.
- That said, JSON arrays are hard to use in a query. As a parser provider, we avoid creating them.